crates/subspace-networking/src/node_runner.rs

```diff
@@ -68,19 +72,18 @@ impl NodeRunner {
             query_id_receivers: HashMap::default(),
             next_subscription_id: 0,
             topic_subscription_senders: HashMap::default(),
+            // We'll make the first query right away and continue at the interval.
+            random_query_timeout: Box::pin(tokio::time::sleep(Duration::from_secs(0)).fuse()),
         }
     }
 
-    pub async fn run(mut self) {
-        // We'll make the first query right away and continue at the interval.
-        let mut random_query_timeout = Box::pin(tokio::time::sleep(Duration::from_secs(0)).fuse());
-
+    pub async fn run(&mut self) {
         loop {
             futures::select! {
-                _ = random_query_timeout => {
+                _ = &mut self.random_query_timeout => {
                     self.handle_random_query_interval();
                     // Increase interval 2x, but to at most 60 seconds.
-                    random_query_timeout = Box::pin(tokio::time::sleep(self.next_random_query_interval).fuse());
+                    self.random_query_timeout = Box::pin(tokio::time::sleep(self.next_random_query_interval).fuse());
                     self.next_random_query_interval = (self.next_random_query_interval * 2).min(Duration::from_secs(60));
                 },
                 swarm_event = self.swarm.next() => {
```

Comment:
Actually, this thing is not resumable still, as normally it just never returns.

Reply:
It is an async function, so it can race against something else that completes sooner and this future will be interrupted in the middle. After that you can call `.run()` again. This is what'll happen when you call `node_runner.run()` in `select!` macro for instance.